src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -65,6 +65,20 @@ impl ScalarFunc {
         Ok(Some(Cow::Owned(format!("{:b}", i).into_bytes())))
     }
 
+    #[inline]
+    pub fn concat<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut output: Vec<u8> = Vec::new();
+        for expr in &self.children {
+            let input = try_opt!(expr.eval_string(ctx, row));
+            output.extend_from_slice(&input);
```

Comment:
I think we need to use `eval_string_and_decode` instead of `eval_string`.

Reply:
@hicqu Binary concat instead of string concat is more accurate, because the decoding may fail. For example, MySQL can concat these stuff anyway:


```
mysql> select hex(concat('123', BINARY convert('测试' using utf32)));
+----------------------------------------------------------+
| hex(concat('123', BINARY convert('测试' using utf32)))   |
+----------------------------------------------------------+
| 31323300006D4B00008BD5                                   |
+----------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select hex(convert('测试' using utf32));
+------------------------------------+
| hex(convert('测试' using utf32))   |
+------------------------------------+
| 00006D4B00008BD5                   |
+------------------------------------+
1 row in set (0.00 sec)
```

Reply:
It's interisting. I have an another test:
```sql
create table x (a varchar(32) charset latin1, b varchar(32));
insert into x values (x'ff', '测试');
select a, b, concat(a,b) as c, hex(concat(a,b)) as d from x;
+------+--------+----------+-------------------+
| a    | b      | c        | hex(concat(a, b)) |
+------+--------+----------+-------------------+
| ÿ    | 测试   | ÿ测试    | C3BFE6B58BE8AF95  |
+------+--------+----------+-------------------+
```
So maybe we should push down both `concat` and `concat_binary`? @zz-jason .

Reply:
@hicqu Actually your SQL is equivalent to:

```
mysql> select concat(convert(convert(x'ff' using latin1) using utf8), '测试');
+-------------------------------------------------------------------+
| concat(convert(convert(x'ff' using latin1) using utf8), '测试')   |
+-------------------------------------------------------------------+
| ÿ测试                                                             |
+-------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select hex(concat(convert(convert(x'ff' using latin1) using utf8), '测试'));
+------------------------------------------------------------------------+
| hex(concat(convert(convert(x'ff' using latin1) using utf8), '测试'))   |
+------------------------------------------------------------------------+
| C3BFE6B58BE8AF95                                                       |
+------------------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select hex(convert(convert(x'ff' using latin1) using utf8));
+------------------------------------------------------+
| hex(convert(convert(x'ff' using latin1) using utf8)) |
+------------------------------------------------------+
| C3BF                                                 |
+------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

Notice that:

```
mysql> select charset(concat(a, b)) from x;
+-----------------------+
| charset(concat(a, b)) |
+-----------------------+
| utf8                  |
+-----------------------+
1 row in set (0.00 sec)
```

But:

```
mysql> select charset(concat(convert(x'ff' using latin1), '测试'));
ERROR 1267 (HY000): Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and (utf8_general_ci,COERCIBLE) for operation 'concat'
```

Simply speaking, MySQL is still doing binary concat (but charset casting rules are different when it is evaluated over table columns or constants).

Reply:
So if the `concat` is pushed down, the internal `convert` will be added by TiDB? If so the implement is good!

Reply:
@zz-jason PTAL Does TiDB's coercible rules needs to be updated to consider string charsets?

Reply:
Yes, we should consider the string charsets during the type inferring phase of the expression functions once we have supported more charsets.

In the execution phase, we just need to concat the string contents, no need to consider the charsets, and set the charsets after concating. 

Take the implementation of MySQL for references: https://github.com/mysql/mysql-server/blob/8.0/sql/item_strfunc.cc#L963

Reply:
Thanks! @zz-jason So according to TiDB and MySQL's implementation, we should just do binary concat here and cast to utf8 (or other charset) is not a task of this function.